pkg/segment/writer/agiletree.go

```diff
@@ -41,6 +42,54 @@ const (
 	TotalMeasFns
 )
 
+var STBHolderPool [MaxConcurrentAgileTrees]*STBHolder
+
+type STBHolder struct {
+	stbPtr            *StarTreeBuilder
+	currentlyInUse    bool
+	lastUsedTimestamp time.Time
+	myIdx             uint16
+}
+
+func GetSTB() *STBHolder {
+	atreeCounterLock.Lock()
+	defer atreeCounterLock.Unlock()
+	if currentAgileTreeCount == MaxConcurrentAgileTrees {
+		return nil
+	}
+
+	for i := 0; i < MaxConcurrentAgileTrees; i++ {
+		if STBHolderPool[i] == nil {
+			STBHolderPool[i] = &STBHolder{}
+			STBHolderPool[i].stbPtr = &StarTreeBuilder{}
+			STBHolderPool[i].currentlyInUse = true
+			STBHolderPool[i].lastUsedTimestamp = time.Now()
+			STBHolderPool[i].myIdx = uint16(i)
+			currentAgileTreeCount++
+
+			return STBHolderPool[i]
+		}
+
+		if !STBHolderPool[i].currentlyInUse {
+			STBHolderPool[i].currentlyInUse = true
+			STBHolderPool[i].lastUsedTimestamp = time.Now()
+			currentAgileTreeCount++
+
+			return STBHolderPool[i]
+		}
```

Comment:
nit: you have quite a bit of duplicate code. I'd suggest one of these options:
1. In the case `if STBHolderPool[i] == nil`, just do the basic setup, and then fall through to `if !STBHolderPool[i].currentlyInUse` to do the code that both parts have
2. Change `var STBHolderPool [MaxConcurrentAgileTrees]*STBHolder` to `var STBHolderPool [MaxConcurrentAgileTrees]STBHolder`, so that you don't need the case `if STBHolderPool[i] == nil` at all. `STBHolder` is already pretty small, so this will only use a little memory (and it'll be on the stack)

Reply:
Went with the first option since we are also allocating StarTreeBuilder when we assign STBHolder and STB has a lot of fields. Addressed in the latest commit.